### PR TITLE
[edn/dv] Rework `edn_err` test to substantially increase coverage

### PR DESCRIPTION
**Pull request by andreaskurth**
This PR reworks the `edn_err` test so that it achieves full coverage of transitions from *each* FSM state to the Error state. This was one of the main coverage gaps left for M2.5.

Please see the commit messages for details. Reviewing this commit by commit is highly recommended to understand the motivation for each change.

### CODE REVIEW

File: hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv [hunk 1]
```diff
@@ -47,18 +48,6 @@ class edn_err_vseq extends edn_base_vseq;
...
-    fork
-      begin
-        // Send INS cmd
-        wr_cmd(.cmd_type("sw"), .acmd(csrng_pkg::INS), .clen(0), .flags(MuBi4False), .glen(0));
-
-        // Send GEN cmd w/ GLEN 1 (request single genbits)
-        wr_cmd(.cmd_type("sw"), .acmd(csrng_pkg::GEN), .clen(0), .flags(MuBi4False), .glen(1));
-      end
-    join_none
-
```
Comment (marnovandermaas) [inline hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv]: Why did you add this for/join and then remove it in the next commit?

File: hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv [hunk 2]
```diff
@@ -16,6 +16,92 @@ class edn_err_vseq extends edn_base_vseq;
...
+      cfg.clk_rst_vif.wait_n_clks(timeout_clks);
+      `uvm_info(`gfn, $sformatf("Timed out waiting for main_sm state %s", exp_state.name()),
+                UVM_HIGH)
+    )
+  endtask
+
+  // Await a randomly selected state of the edn_ack_sm at index `idx`.  If the state is not reached
+  // within `timeout_clks` clock cycles, the task also returns.  This task returns immediately with
+  // a chance of 100 - `await_pct` percent.
+  task await_random_ack_sm_state(int timeout_clks = 1_000, int await_pct = 90);
```
Comment (marnovandermaas) [inline hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv:71]: Some duplicated code with `await_random_main_sm_state` is there any simple way to de-duplicate this?

**Review by marnovandermaas — COMMENTED** [review]
Thanks for these changes, a couple of small comments from my side.

File: hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv (hunk 1 above)
Comment (andreaskurth) [reply hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv]: In the second commit I added a fork-join around the existing `wr_cmd` to move them from a foreground thread to a background thread.

In the third commit I changed the `wr_cmd`s to generalize them from SW-mode-only to Auto mode, Boot mode, and SW mode. In the same commit, I moved the code a few lines down, which may be confusing and is no longer necessary (I originally had added more code between).

If you prefer, I can modify the third commit so that the generalized `wr_cmd`s are on the same location as the SW-mode-only ones were before the third commit.

File: hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv (hunk 2 above)
Comment (andreaskurth) [reply hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv:71]: True, thanks for the note. I think moving the `DV_SPINWAIT` code into a common task can reduce duplication

File: hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv (hunk 1 above)
Comment (marnovandermaas) [reply hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv]: Ah, now I see the code below. No, it's fine this way. I was just confused earlier and this clarifies it.

File: hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv (hunk 2 above)
Comment (andreaskurth) [reply hw/ip/edn/dv/env/seq_lib/edn_err_vseq.sv:71]: I've tried refactoring it, but `uvm_hdl_read` returns a `uvm_hdl_data_t`, which is too wide, so it needs to be reduced with a cast or bit select to the type of the expected value in the comparison. SV does not support a bit select of dynamic width, though. I don't know it's possible to pass a type argument into a task (I didn't try it), but I gave up after spending 30 min or so on this -- not worth the effort given prios for a bit of code deduplication.

**Comment by andreaskurth** [comment]
Merged as the window for implementing changes for me is closing. Happy to address changes next week if necessary.
